### PR TITLE
ENH: support SpooledTemporaryFile

### LINKED ISSUE #44748: BUG: pd.read_csv - encoding argument ignored for SpooledTemporaryFile
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
import tempfile
f = tempfile.SpooledTemporaryFile()
f.write('''\
beløp,valuta
10.0,BTC
5.0,ETH
'''.encode('latin-1'))
f.seek(0)
pd.read_csv(f, encoding='latin-1')
# UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf8 in position 3: invalid start byte

# If I switch SpooledTemporaryFile with TemporaryFile, everything works
```

### Issue Description

When passing a SpooledTemporaryFile to pd.read_csv, the encoding-argument is ignored, and it seems like it tries to parse the bytes as utf-8. This will fail if the bytes contain non utf-8 characters and uses a different encoding.

In the example above the text contains a Norwegian letter in latin-1 encoding. Note that if I switch from using a SpooledTemporaryFile to a TemporaryFile, the problem goes away, and pd.read_csv uses the correct encoding.

### Expected Behavior

pd.read_csv should respect the encoding argument when receiving a SpooledTemporaryFile, and be able to correctly parse the example above without raising an exception.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 73c68257545b5f8530b7044f56647bd2db92e2ba
python           : 3.8.12.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.6.0
Version          : Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8

pandas           : 1.3.3
numpy            : 1.21.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.1.3
setuptools       : 57.2.0.post0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.28.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by twoertwein**
- [x] closes #44748
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/parser/test_encoding.py [hunk 1]
```diff
@@ -302,13 +302,17 @@ def test_readcsv_memmap_utf8(all_parsers):
...
+@pytest.mark.parametrize("mode", ["w+b", "w+t"])
+def test_not_readable(all_parsers, mode):
     # GH43439
     parser = all_parsers
-    if parser.engine in ("python", "pyarrow"):
-        pytest.skip("SpooledTemporaryFile does only work with the c-engine")
-    with tempfile.SpooledTemporaryFile() as handle:
-        handle.write(b"abcd")
+    if parser.engine == "pyarrow":
+        pytest.skip("SpooledTemporaryFile does only work with the pyarrow-engine")
```
Comment (jbrockmendel) [inline pandas/tests/io/parser/test_encoding.py:310]: can you make this an xfail?  ive been finding many of our skips have been silently fixed

File: pandas/io/common.py [hunk 2]
```diff
@@ -935,7 +930,7 @@ def __init__(
         self.decode = decode
 
         self.attributes = {}
-        for attribute in ("seekable", "readable", "writeable"):
```
Comment (twoertwein) [inline pandas/io/common.py:938]: Should have been `"writable"` but it seems it was never used as we do not support writing to memory-mapped files.

**Comment by twoertwein** [comment]
A trimmed down version of this PR could go to 1.3.* as it technically fixes a regression. Let me know whether it should be back-ported or not, based on that I will create a whatsnew for either 1.3 or 1.4.

**Comment by jreback** [comment]
let's just put in 1.4, trying to avoid backporting anything unless its critical

File: pandas/tests/io/parser/test_encoding.py [hunk 3]
```diff
@@ -299,17 +299,21 @@ def test_readcsv_memmap_utf8(all_parsers):
...
+@pytest.mark.usefixtures("pyarrow_xfail")
+@pytest.mark.parametrize("mode", ["w+b", "w+t"])
+def test_not_readable(all_parsers, mode, request):
     # GH43439
     parser = all_parsers
-    if parser.engine in ("python", "pyarrow"):
+    if parser.engine == "pyarrow":
         mark = pytest.mark.xfail(
-            reason="SpooledTemporaryFile does only work with the c-engine"
+            reason="SpooledTemporaryFile does only work with the pyarrow-engine"
```
Comment (jreback) [inline pandas/tests/io/parser/test_encoding.py:309]: umm is this reason correct? you mean does *not* right?

File: pandas/io/common.py [hunk 4]
```diff
@@ -976,11 +971,22 @@ def __next__(self) -> str:
         return newline.lstrip("\n")
 
 
-# Wrapper that wraps a StringIO buffer and reads bytes from it
-# Created for compat with pyarrow read_csv
-class BytesIOWrapper(io.BytesIO):
-    buffer: StringIO | TextIOBase | None
+class _IOWrapper:
+    # Lies that certain attributes exist and are True.
```
Comment (jbrockmendel) [inline pandas/io/common.py:975]: this comment makes sense in context, but might merit fleshing out for future (or non-native-English) readers

File: pandas/tests/io/parser/test_encoding.py (hunk 3 above)
Comment (twoertwein) [reply pandas/tests/io/parser/test_encoding.py]: thank you, fixed it :)

File: pandas/io/common.py (hunk 4 above)
Comment (twoertwein) [reply pandas/io/common.py]: I wrote a more lengthy comment

File: pandas/tests/io/parser/test_encoding.py [hunk 5]
```diff
@@ -299,17 +299,21 @@ def test_readcsv_memmap_utf8(all_parsers):
     tm.assert_frame_equal(df, dfr)
 
 
-def test_not_readable(all_parsers, request):
+@pytest.mark.usefixtures("pyarrow_xfail")
```
Comment (twoertwein) [inline pandas/tests/io/parser/test_encoding.py:302]: @jbrockmendel  Is this needed? Already have  `pytest.mark.xfail` in the body.

Comment (jbrockmendel) [reply pandas/tests/io/parser/test_encoding.py:302]: i think either works, but only need one.

**Comment by twoertwein** [comment]
Thank you for rebasing @jreback, green except for (unrelated?) codecov error.

File: pandas/io/common.py [hunk 6]
```diff
@@ -976,11 +971,30 @@ def __next__(self) -> str:
...
+class _IOWrapper:
+    # TextIOWrapper is overly strict: it request that the buffer has seekable, readable,
+    # and writable. If we have a read-only buffer, we shouldn't need writable and vice
+    # versa. Some buffers, are seek/read/writ-able but they do not have the "-able"
+    # methods, e.g., tempfile.SpooledTemporaryFile.
+    # If a buffer does not have the above "-able" methods, we simple assume they are
+    # seek/read/writ-able.
+    def __init__(self, buffer: BaseBuffer):
+        self.buffer = buffer
+        self.attributes = tuple(
```
Comment (jreback) [inline pandas/io/common.py:983]: can you make these separate properties and just do the check of the buffer inside them? its a bit longer code but should be more obvious what is going on

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/common.py (hunk 6 above)
Comment (twoertwein) [reply pandas/io/common.py]: Okay, changed it. I made them methods and not properties, as `typing.IO` defines them as methods:
https://github.com/python/typeshed/blob/1d5857e1e0e0fe6078e563d769f07fe3a3352cc9/stdlib/typing.pyi#L522

Comment (jreback) [reply pandas/io/common.py]: oh sure

**Comment by jreback** [comment]
thanks @twoertwein this looks great (yeah a bit longer but way more readable). waiting to merge as i hope some CI patches get us back to green and then can rebase here.

**Comment by jreback** [comment]
are the windows checks related?

**Comment by twoertwein** [comment]
It should be unrelated (the test case doesn't seem to open any files). Will rebase and force push to test it again. Cannot reproduce it on linux.

**Comment by twoertwein** [comment]
The error disappeared but I can imagine that it might be related: if mmap fails, opened handles will not be released

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @twoertwein

**Pull request merged**